### PR TITLE
Tool to create a psibase project from template

### PR DESCRIPTION
**Pull request by cool-ant**
Template that can be used via `cargo-generate` to instantiate a psibase project.

### CODE REVIEW

**Review by Velua — APPROVED** [review]

File: service-templates/rust/plugin/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+use crate::bindings::host::common::types::{Error, PluginId};
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum ErrorType {
+    QueryResponseParseError,
+}
```
Comment (swatanabe) [inline service-templates/rust/plugin/src/errors.rs:6]: Please use the `plugin_error!` macro to define errors.

File: service-templates/rust/plugin/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    fn set_example_thing(thing: String) {
+        let packed_example_thing_args = {{project-name | snake_case}}::action_structs::setExampleThing { thing }.packed();
+        add_action_to_transaction("setExampleThing", &packed_example_thing_args).unwrap();
+    }
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct ExampleThingData {
+    exampleThing: String,
```
Comment (swatanabe) [inline service-templates/rust/plugin/src/lib.rs:27]: If you're using `rename_all`, the member name should be `example_thing`.

File: service-templates/rust/plugin/src/errors.rs (hunk 1 above)
Comment (cool-ant) [reply service-templates/rust/plugin/src/errors.rs]: hm, must have found the one place to pull it from you didn't already replace. PR to handle the one isntance I found: #955

File: service-templates/rust/ui/src/shad/components/ui/dropdown-menu.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
+import * as React from "react";
```
Comment (brandonfancher) [inline service-templates/rust/ui/src/shad/components/ui/dropdown-menu.tsx:1]: This file appears to be triplicated. Once here, and again in `src/components` and `src/components/ui`.

File: service-templates/rust/ui/src/shad/components/ui/button.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import * as React from "react";
```
Comment (brandonfancher) [inline service-templates/rust/ui/src/shad/components/ui/button.tsx:1]: This file appears to be duplicated. Once here, and again in `src/components/ui`.

File: service-templates/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+# Usage
+
+## Instantiating the template
+
+> Only for Rust projects so far
+
+Instanstiate a Rust service from Git:
+
+`cargo generate --git fractally/psibase/templates` rust
```
Comment (swatanabe) [inline service-templates/README.md:15]: I don't see how this can work.

File: service-templates/rust/plugin/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+impl Queries for {{project-name | upper_camel_case}}Plugin {
+    fn get_example_thing() -> Result<String, Error> {
+        let graphql_str = "query { exampleThing }";
+
+        let examplething_val = serde_json::from_str::<ExampleThingResponse>(
+            &CommonServer::post_graphql_get_json(&graphql_str)?,
+        );
+
+        let examplething_val = examplething_val
+            .map_err(|err| ErrorType::QueryResponseParseError.err(err.to_string().as_str()))?;
```
Comment (swatanabe) [inline service-templates/rust/plugin/src/lib.rs:44]: Did you check that the generated code compiles?

File: service-templates/rust/ui/index.html [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Psibase Templated UI</title>
```
Comment (swatanabe) [inline service-templates/rust/ui/index.html:7]: Can this be the app name converted to Title Case.

Comment (cool-ant) [reply service-templates/rust/ui/index.html:7]: absolutely. good call

File: service-templates/README.md
Comment (James-Mart) [inline service-templates/README.md:1]: A more accurate name for the directory would be "package-templates"

File: service-templates/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+Instanstiate a Rust service from local path in `services` directory.
+
+`cargo generate -p ../../service-templates/rust`
+
+## Building a package
+
+1. Build the UI by running `yarn` and `yarn build` in the `ui/` directory.
```
Comment (James-Mart) [inline service-templates/README.md:23]: When I yarn build I get 
```
src/App.tsx:8:21 - error TS2307: Cannot find module '@components/nav' or its corresponding type declarations.
8 import { Nav } from "@components/nav";
```

File: service-templates/rust/plugin/src/lib.rs (hunk 6 above)
Comment (cool-ant) [reply service-templates/rust/plugin/src/lib.rs]: I did, but I've gotta test the instantiation and update the template; looks like I missed copying this change.

File: service-templates/README.md (hunk 5 above)
Comment (swatanabe) [reply service-templates/README.md]: The other problem: our repo is gofractally

File: service-templates/README.md
Comment (swatanabe) [reply service-templates/README.md:1]: Or maybe "app-templates"

File: service-templates/README.md (hunk 5 above)
Comment (cool-ant) [reply service-templates/README.md]: Doh! Right, that's bad. Thanks for the safety tip Egon!
image

File: package-templates/README.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+> Only for Rust projects so far
+
+Instanstiate a Rust service from Git:
+
+`cargo generate --git gofractally/psibase` package-templates/rust
+
+Instanstiate a Rust service from local path in `services` directory.
+
+`cargo generate -p ../../service-templates/rust`
```
Comment (James-Mart) [inline package-templates/README.md:19]: directory was renamed

File: package-templates/rust/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,2 @@
+pub use {{project-name}};
+pub use {{project-name}}-plugin;
```
Comment (James-Mart) [inline package-templates/rust/src/lib.rs:2]: This `pub use *-plugin` line should be removed. I've seen this in other packages, but it's wrong in those, too.

File: package-templates/rust/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,2 @@
+pub use {{project-name}};
```
Comment (James-Mart) [inline package-templates/rust/src/lib.rs:1]: This should be 
`pub use {{project-name | snake_case}}`

**Review by James-Mart — CHANGES_REQUESTED** [review]

File: package-templates/README.md (hunk 9 above)
Comment (cool-ant) [reply package-templates/README.md]: 🤦

**Review by James-Mart — APPROVED** [review]

File: package-templates/rust/ui/src/components/nav.tsx [hunk 12]
```diff
@@ -0,0 +1,36 @@
+import { HoverBorderGradient } from "@shadcn/hover-border-gradient";
```
Comment (James-Mart) [inline package-templates/rust/ui/src/components/nav.tsx:1]: `error TS2307: Cannot find module '@shadcn/hover-border-gradient' or its corresponding type declarations.`

**Review by James-Mart — CHANGES_REQUESTED** [review]

**Review by James-Mart — APPROVED** [review]

**Pull request merged**